src/BalanceTx/ExUnitsAndMinFee.purs

```diff
@@ -60,22 +65,34 @@ import Types.Scripts (PlutusScript)
 import Types.Transaction (TransactionInput)
 import Types.UnbalancedTransaction (_transaction)
 import Untagged.Union (asOneOf)
+import TxOutput (transactionInputToTxOutRef, transactionOutputToOgmiosTxOut)
 
 evalTxExecutionUnits
   :: Transaction
   -> UnattachedUnbalancedTx
+  -> UtxoMap
   -> BalanceTxM Ogmios.TxEvaluationResult
-evalTxExecutionUnits tx unattachedTx = do
+evalTxExecutionUnits tx unattachedTx utxos = do
   txBytes <- liftEffect
     ( wrap <<< Serialization.toBytes <<< asOneOf <$>
         Serialization.convertTransaction tx
     )
-  eResult <- unwrap <$> lift (QueryM.evaluateTxOgmios txBytes)
+  eResult <- unwrap <$> lift (QueryM.evaluateTxOgmios txBytes additionalUtxoSet)
   case eResult of
     Right a -> pure a
     Left e | tx ^. _isValid -> throwError $ ExUnitsEvaluationFailed unattachedTx
       e
     Left _ -> pure $ wrap $ Map.empty
+  where
+  additionalUtxoSet :: Ogmios.AdditionalUtxoSet
+  additionalUtxoSet = Ogmios.AdditionalUtxoSet $ Map.fromFoldable utxos'
+
+  utxos' :: Array (Ogmios.OgmiosTxOutRef /\ Ogmios.OgmiosTxOut)
+  utxos' =
+    ( \(inp /\ out) ->
+        (transactionInputToTxOutRef inp /\ transactionOutputToOgmiosTxOut out)
+    )
```

Comment:
We can use [bimap](https://pursuit.purescript.org/packages/purescript-bifunctors/6.0.0/docs/Data.Bifunctor#v:bimap) here. I would also merge `additionalUtxoSet` and `utxos'` into a single function, since `utxos'` isn't used anywhere else.